python/ray/air/preprocessors/stateful_preprocessor.py

```diff
@@ -0,0 +1,120 @@
+from typing import Callable, TYPE_CHECKING, Dict
+
+from ray.air.preprocessor import Preprocessor
+from ray.data import Dataset
+
+if TYPE_CHECKING:
+    import pandas
+
+
+class StatefulPreprocessor(Preprocessor):
```

Comment:
Should our existing preprocessors be using this interface?

Reply:
Yeah this is a great question. I asked myself this while working on this PR and had the same logic as Antoni, I've updated the PR description with this. We can definitely revisit this and refactor the existing Preprocessors down the line if we do decide to converge on this.